File: 2022/day7/day7.py
from typing import List, Dict
from dataclasses import dataclass
import logging

logging.basicConfig(level=logging.DEBUG)


@dataclass
class File:
    size: int
    name: str  # not needed

    def __repr__(self):
        return f"{self.name} (file, size={self.size})"


@dataclass
class Directory:
    name: str
    parent: "Directory"
    files: List[File]
    children: Dict

    def size(self):
        return sum(file.size for file in self.files) + sum(
            child.size() for child in self.children.values()
        )

    def addFile(self, file):
        self.files.append(file)

    def addChild(self, name, directory):
        if name in self.children:
            raise ValueError(f"{name} already present")
        self.children[name] = directory

    def __repr__(self):
        output = "\n"
        for file in self.files:
            output += f"{file}\n"
        for child in self.children.values():
            output += f"{child}\n"
        return output


def parse_input(instructions):
    root = Directory(name="/", parent=None, files=[], children={})
    node = root
    i = 1
    while i < len(instructions):
        line = instructions[i]
        assert line.startswith("$")
        # could be cd or ls
        if "ls" in line.split():
            # ls: add files and children to node
            i += 1
            line = instructions[i]
            while not line.startswith("$"):
                if line.startswith("dir"):
                    # add a directory to node
                    name = line.split()[-1]
                    directory = Directory(name=name, files=[], children={}, parent=node)
                    logging.debug("adding a child directory %s to %s", name, node.name)
                    node.addChild(
                        name=name,
                        directory=directory,
                    )
                else:
                    # add a file to node
                    size, name = line.split()
                    file = File(size=int(size), name=name)
                    logging.debug("adding a file %s to %s", file, node.name)
                    node.addFile(file)
                # go on to the next line
                i += 1
                try:
                    line = instructions[i]
                except IndexError:
                    # reached end of file
                    break
        else:
            # cd: change node
            target = line.split()[-1]
            if target == "..":
                node = node.parent
            else:
                # change into a subdirectory
                logging.debug("Moving into %s", node.name + "/" + target)
                node = node.children[target]
            i += 1
    return root


def add_up_sizes(directory, threshold=100_000):
    size = directory.size()
    if size > threshold:
        size = 0
    return size + sum(
        add_up_sizes(child, threshold) for child in directory.children.values()
    )


if __name__ == "__main__":
    fname = "sample.txt"  # sys.argv[1]
    with open(fname, "r") as f:
        instructions = f.readlines()
    root = parse_input(instructions)
    assert root.size() == 48381165
    assert root.children["d"].size() == 24933642
    assert root.children["a"].children["e"].size() == 584
    assert root.children["a"].size() == 94853
    total_size = add_up_sizes(root, 100_000)
    assert total_size == 95437, total_size

    # part 1
    fname = "input.txt"
    with open(fname, "r") as f:
        instructions = f.readlines()
    root = parse_input(instructions)
    print(add_up_sizes(root, 100_000))

File: 2022/day7/day7.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Directory struct {
	name     string
	parent   *Directory
	children []*Directory
	files    int
}

func (dir Directory) sizes() []int {
	sizes := make([]int, 1)
	sizes[0] = dir.size()
	for _, child := range dir.children {
		for _, size := range child.sizes() {
			sizes = append(sizes, size)
		}
	}
	return sizes
}

func (dir Directory) size() int {
	size := dir.files
	for _, child := range dir.children {
		size += child.size()
	}
	return size
}

func (dir *Directory) addChild(child *Directory) {
	dir.children = append(dir.children, child)
}

func (dir *Directory) addFile(filesize int) {
	dir.files += filesize
}

func isCommand(line string) bool {
	return line[0:1] == "$"
}

func isDirectory(line string) bool {
	return line[:3] == "dir"
}

func isCD(line string) bool {
	return line[:4] == "$ cd"
}

func isLS(line string) bool {
	return line[:4] == "$ ls"
}

func day7(filename string, part1 bool) int {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var root Directory
	var currentDir *Directory // currentDir is a pointer
	root = Directory{name: "/", parent: nil, children: make([]*Directory, 0), files: 0}
	currentDir = &root
	lines := make([]string, 0)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	// fmt.Println(strings.Join(lines, "\n")) . // same as input
	// skip the first one, which just does cd into /
	var i int
	for i = 1; i < len(lines); i++ {
		fmt.Println("current working directory is `" + currentDir.name + "`")
		if isCD(lines[i]) {
			destination := strings.Split(lines[i], " ")[2]
			fmt.Println("I want to change working directory to " + destination)
			if destination == ".." {
				currentDir = currentDir.parent
			} else {
				for _, child := range currentDir.children {
					if child.name == destination {
						currentDir = child
						fmt.Println("I changed working directory to " + currentDir.name)
						break
					} else {
						fmt.Println(child.name + " != " + destination)
					}
				}
				// msg := "could not find subdirectory " + destination + " of parent " + currentDir.name + ". Its children are ("
				// for _, child := range currentDir.children {
				// 	msg += child.name + ","
				// }
				// msg += ")"
				// panic(msg)

			}

		} else if isLS(lines[i]) {
			// Need to look at the next lines
			for i = i + 1; i < len(lines) && !isCommand(lines[i]); i++ {
				if isDirectory(lines[i]) {
					child := Directory{name: strings.Split(lines[i], " ")[1],
						parent: currentDir, children: make([]*Directory, 0),
						files: 0}
					currentDir.addChild(&child)
					fmt.Println("I added directory " + child.name + " to " + currentDir.name)
				} else {
					// it's a file
					oldSize := int64(currentDir.files)
					fileName := strings.Split(lines[i], " ")[1]
					fileSize, _ := strconv.Atoi(strings.Split(lines[i], " ")[0])
					currentDir.addFile(fileSize)
					newSize := int64(currentDir.files)
					fmt.Println("I added file " + fileName + " to " + currentDir.name)
					fmt.Println("its size increased from " + strconv.FormatInt(oldSize, 10) +
						" to " + strconv.FormatInt(newSize, 10))
				}
			}
			// fmt.Println("I broke out of the ls loop. i=" + strconv.FormatInt(int64(i), 10))
			i--

		} else {
			panic("Unknown command" + lines[i])
		}

	}
	sizes := root.sizes()
	// part 1
	total := 0
	for _, size := range sizes {
		if size < 100000 {
			total += size
		}
	}
	if part1 {
		return total
	}
	// part 2
	freeSpace := 70000000 - root.size()
	requiredFreeSpace := 30000000
	minSoFar := root.size()
	for _, size := range sizes {
		newFreeSpace := freeSpace + size
		if newFreeSpace >= requiredFreeSpace {
			if size < minSoFar {
				minSoFar = size
			}
		}
	}
	return minSoFar
}

func main() {
	// part 1
	if day7("sample.txt", true) != 95437 {
		panic("wrong answer to sample")
	}
	fmt.Println("the answer to part1 is ")
	fmt.Println(day7("input.txt", true))
	// part 2
	if day7("sample.txt", false) != 24933642 {
		panic("Wrong answer to sample problem, part 2")
	}
	fmt.Println("the answer to part2 is ")
	fmt.Println(day7("input.txt", false))
}
